### ISSUE TITLE
Rotated bounding box iou returns negative value

### ISSUE THREAD

**Issue by edwardchaos** [facebookresearch/detectron2#350]
During training of rotated bounding box detection, matcher [at line 82](https://github.com/facebookresearch/detectron2/blob/f485b4684c5cf58b53263706be386f57e28d1841/detectron2/modeling/matcher.py#L82) sometimes fails.

Match quality is produced by rotated bounding box iou extension in files detectron2/layers/csrc/box_iou_rotated/. The output values of iou should always be in range [0, 1]

Specifically, the function single_box_iou_rotated in box_iou_rotated_utils.h isn't guaranteed to be in range [0, 1].

Workaround is to clamp values in match quality matrix in range [0, 1] so it passes the assertion

Perhaps add checks for valid input bounding box and assertion for valid output.

## Environment

------------------------  --------------------------------------------------
sys.platform              linux
Python                    3.7.5 (default, Oct 25 2019, 15:51:11) [GCC 7.3.0]
Numpy                     1.17.3
Detectron2 Compiler       GCC 7.4
Detectron2 CUDA Compiler  10.1
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.3.1
PyTorch Debug Build       False
torchvision               0.4.2
CUDA available            True
GPU 0,1                   GeForce RTX 2080 Ti
CUDA_HOME                 /usr/local/cuda-10.1
NVCC                      Cuda compilation tools, release 10.1, V10.1.243
Pillow                    6.2.1
cv2                       4.1.1
------------------------  --------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

**Comment by ppwwyyxx** [comment]
Could you provide some sample data that causes iou to return negative values?

**Comment by ppwwyyxx** [comment]
closing as there is no response.
cc @SuperIRabbit in case there are any numerical issues you can think of about iou.

**Comment by SuperIRabbit** [comment]
@edwardchaos It would be great if you could provide sample data that cause the issue, thanks!

**Comment by edwardchaos** [comment]
@SuperIRabbit yes, i'll try to find some time this week

**Comment by edwardchaos** [comment]
trained overnight, 11660 iterations after removing the torch.clamp guard. Cannot reproduce error. Let's consider it a user error at this point.

**Comment by dnandha** [comment]
having the same problem!! narrowed down the problem a bit: cuda kernel receives boxes with either zero width or zero height, but the GT's are > 0 and right before calling `pairwise_iou` all widths and heights are > 0. reopening issue soon.

**Comment by Genie-Kim** [comment]
same problem here!!
I am using rotated bbox, when I call pairwise_iou_rotated() in rotated_fast_rcnn.py, there are many negative value in match_quality_matrix that is output of pairwise_iou_rotated().

**Comment by cbasavaraj** [comment]
Hello, I have the same issue, not often, but happens at random. Trying to debug. Will post if I find something.

**Comment by dnandha** [comment]
@Genie-Kim @cbasavaraj I have debugged my way through the code and found the solution. Without further ado, to fix the problem add following option to your configuration:
`MODEL:
  PROPOSAL_GENERATOR:
    MIN_SIZE: 1`
Also opened up PR #1076 for change of default config.

**Comment by cbasavaraj** [comment]
Thanks @dnandha ! Looks like @facebook-github-bot requires you to sign a CLA (see your pull-request).

Do you know if the unit for this `MIN_SIZE`? Pixels? Or stride-length? The documentation is not clear:
```
# Proposal height and width both need to be greater than MIN_SIZE
# (a the scale used during training or inference)
_C.MODEL.PROPOSAL_GENERATOR.MIN_SIZE = 0
```

**Comment by cbasavaraj** [comment]
Another question @dnandha . Were you able to get good results with rotated boxes? I'm still training, but so far losses and average precisions are worse than what I had with axis-aligned boxes? I'm beginning to think predicting the angle may be too tough? 

However, I read this paper yesterday: https://arxiv.org/abs/1703.01086, which is essentially Faster RCNN (for text) with RRPN and RROI pooling. They match ground-truth boxes with anchors using not just IoU overlap, but also require anchor theta to be with 15 degrees of GT theta. I think that may be worth trying.
